Log correct caller line number and more control over verbosity

prydie opened the pull request:
I've made the following changes for inclusion in another project and thought I'd see if there was interest in upstreaming them:

 - Log the correct caller line number: Previously we logged the line number in this wrapper rather than the line number of the log statement in calling code.
 - Report verbosity based on both `-v` and zap lvl
   * Display `glog.V(n)` `n <= 2` if `zap.InfoLevel` is enabled.
   * Display `glog.V(n)` `n <= 4` if `zap.DebugLevel` is enabled.
   * Display of `glog.V(n)` `n > 4` now controlled by -v=n flag.
   * NOTE: The package now declares the `-v` flag which is a breaking change if importing projects already declare the flag themselves, however, fits within the glog package interface.

I also added in dependency management via `dep` but that may or may not be desirable.

prydie committed:
Manage dependencies with dep

prydie committed:
Log the correct caller line number

Previously we logged the line number in this wrapper rather than the
line number of the log statement in calling code.

prydie committed:
Report verbosity based on both -v and zap lvl

 * Display glog.V(n) n <= 2 if zap.InfoLevel is enabled.
 * Display glog.V(n) n <= 4 if zap.DebugLevel is enabled.
 * Display of glog.V(n) n > 4 now controlled by -v=n flag.

istio-testing commented:
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by:
To fully approve this pull request, please assign additional approvers.
We suggest the following additional approvers: 

If they are not already assigned, you can assign the PR to them by writing `/assign ` in a comment when ready.

The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details open>
Needs approval from an approver in each of these files:

- **[OWNERS](https://github.com/istio/glog/blob/master/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":[]} -->

prydie commented:
/assign @geeknoid

geeknoid commented on glog.go:
```diff
@@ -38,136 +98,167 @@ func Flush() {
+// skipLogger constructs a suggared logger with an additional caller skip to
+// ensure we log the correct line number.
+func skipLogger() *zap.SugaredLogger {
+	return zap.L().WithOptions(zap.AddCallerSkip(1)).Sugar()
```
Could this be allocated and reused? As is, this could cause a lot of garbage.

prydie replied:
It could be allocated in `func int()` but callers would need to swap in their own configured logger explicitly via an analogue of [`zap.ReplaceGlobals()`](https://godoc.org/go.uber.org/zap#ReplaceGlobals) which would mean the package was no longer a complete drop-in replacement for glog.

With this implementation, the user can just call [`zap.ReplaceGlobals()`](https://godoc.org/go.uber.org/zap#ReplaceGlobals) directly and we'll use their configured logger.

On reflection maybe the performance improvement is worth the tradeoff?

geeknoid commented:
Instead of doing it in init, can you have something that just does:

  read global logger
  if global logger has changed
      allocate a new logger object


So basically, cache the entry until the global logger changes, and then
realloc, etc.

On Mon, Aug 13, 2018 at 9:57 AM Andrew Pryde <notifications@github.com>
wrote:

> *@prydie* commented on this pull request.
> ------------------------------
>
> In glog.go <https://github.com/istio/glog/pull/2#discussion_r209683999>:
>
> > @@ -38,136 +98,167 @@ func Flush() {
>  	zap.L().Sync()
>  }
>
> -// V is a shim
> +// skipLogger constructs a suggared logger with an additional caller skip to
> +// ensure we log the correct line number.
> +func skipLogger() *zap.SugaredLogger {
> +	return zap.L().WithOptions(zap.AddCallerSkip(1)).Sugar()
>
> It could be allocated in func int() but callers would need to swap in
> their own configured logger explicitly via an analogue of
> zap.ReplaceGlobals() <https://godoc.org/go.uber.org/zap#ReplaceGlobals>
> which would mean the package was no longer a complete drop-in replacement
> for glog.
>
> With this implementation, the user can just call zap.ReplaceGlobals()
> <https://godoc.org/go.uber.org/zap#ReplaceGlobals> directly and we'll use
> their configured logger.
>
> On reflection maybe the performance improvement is worth the tradeoff?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/istio/glog/pull/2#discussion_r209683999>, or mute the
> thread
> <https://github.com/notifications/unsubscribe-auth/AVucHUEA12Wi9B1nYeOdeym5PXr6nSzEks5uQbAIgaJpZM4V6ZVc>
> .
>

prydie committed:
Add cache to skipLogger to improve performance

Before:
BenchmarkSkipLogger-4           10000000               175 ns/op

After:
BenchmarkSkipLogger-4           50000000                36.8 ns/op

prydie commented:
@geeknoid hopefully the new implementation is better (and correct!):


**Before**
```
BenchmarkSkipLogger-4           10000000               175 ns/op
```
**After**
```
BenchmarkSkipLogger-4           50000000                36.8 ns/op
```

**Only calling** `zap.S()`
```
BenchmarkSkipLogger3-4          100000000               18.7 ns/op
```

prydie committed:
Unit test for zap.ReplaceGlobals()

geeknoid commented on glog.go:
```diff
@@ -101,7 +107,17 @@ func Flush() {
+		defer logging.mu.Unlock()
+		logging.logger = zap.L().WithOptions(zap.AddCallerSkip(1)).Sugar()
+		logging.zapGlobal = zap.L()
+	}
```
Add cached := logging.logger

geeknoid commented on glog.go:
```diff
@@ -101,7 +107,17 @@ func Flush() {
+		logging.logger = zap.L().WithOptions(zap.AddCallerSkip(1)).Sugar()
+		logging.zapGlobal = zap.L()
+	}
+	return logging.logger
```
change to "return cached".

Otherwise, the race detector will complain that logging.logger is accessed in a racy way.

prydie committed:
Code review comments

prydie commented:
@geeknoid Like so?

geeknoid commented on glog.go:
```diff
@@ -38,136 +104,180 @@ func Flush() {
+		logging.zapGlobal = zap.L()
+	}
+
+	cached := logging.logger
```
Nope, not quite what I meant. logging.logger needs to only be read while the mutex is held. So something like:

func skipLogger() *zap.SugaredLogger {
	logging.mu.RLock()
	cachedGlobal := logging.zapGlobal
	cachedLogger := logging.logger
	logging.mu.RUnlock()

	global := zap.L()
	if cachedGlobal !== global {
		logging.mu.Lock()
		logging.zapGlobal = global
		logging.logger = global.WithOptions(zap.AddCallerSkip(1)).Sugar()
		cachedLogger = logging.logger
		logging.mu.Unlock()
	}

	return cachedLogger;
}

prydie commented:
@geeknoid Ah! Of course. Should be good now 🙂

geeknoid commented:
Please reopen if you have time to address the feedback.
